### ISSUE TITLE
Module that has dependency on its own external ID cannot be installed

### ISSUE THREAD

**Issue by AdrianTeng** [odoo/odoo#83385]
### Impacted versions: 
13.0

### Steps to reproduce:

**TL; DR**: _My module has a new model, some new rows on that model, and an existing model that has many2one field to this new model with default values on those new rows. Odoo seems to wanting to resolve the default value before the xml data and I got error on `External ID not found` during installation._

Given a custom minimalist module `test_install_order` with following content:

```bash
test_install_order
├── __init__.py
├── __manifest__.py
├── data
│   └── some_data.xml
└── models
    ├── __init__.py
    └── product.py
```

(Omitting non-essential details)

`product.py`

```
from odoo import fields, models

class ProductCustomTag(models.Model):
    _name = 'product.custom.tag'

    name = fields.Char()


class ProductTemplate(models.Model):
    _inherit = 'product.template'

    custom_tag_id = fields.Many2one('product.custom.tag', default=lambda self: self.env.ref('test_install_order.default_tag'))
```

`some_data.xml`

```
<?xml version="1.0" encoding="utf-8" ?>
<odoo>
    <record model="product.custom.tag" id="default_tag">
        <field name="name">Default</field>
    </record>
</odoo>
```

And finally I've ensured the relevant `__init__.py` and `__manifest__.py` have referenced the python / xml files respectively.

With this addon, if one attempt to install (not update!) the addon, we will get the following error:

```
Error:
Odoo Server Error

Traceback (most recent call last):
  File "/Users/adrian/odoo13/odoo/odoo/tools/cache.py", line 85, in lookup
    r = d[key]
  File "/Users/adrian/odoo13/odoo/odoo/tools/func.py", line 69, in wrapper
    return func(self, *args, **kwargs)
  File "/Users/adrian/odoo13/odoo/odoo/tools/lru.py", line 44, in __getitem__
    a = self.d[obj].me
KeyError: ('ir.model.data', <function IrModelData.xmlid_lookup at 0x1117db040>, 'test_install_order.default_tag')

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/adrian/odoo13/odoo/odoo/http.py", line 624, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/Users/adrian/odoo13/odoo/odoo/http.py", line 310, in _handle_exception
    raise pycompat.reraise(type(exception), exception, sys.exc_info()[2])
  File "/Users/adrian/odoo13/odoo/odoo/tools/pycompat.py", line 14, in reraise
    raise value
  File "/Users/adrian/odoo13/odoo/odoo/http.py", line 669, in dispatch
    result = self._call_function(**self.params)
  File "/Users/adrian/odoo13/odoo/odoo/http.py", line 350, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/Users/adrian/odoo13/odoo/odoo/service/model.py", line 94, in wrapper
    return f(dbname, *args, **kwargs)
  File "/Users/adrian/odoo13/odoo/odoo/http.py", line 339, in checked_call
    result = self.endpoint(*a, **kw)
  File "/Users/adrian/odoo13/odoo/odoo/http.py", line 915, in __call__
    return self.method(*args, **kw)
  File "/Users/adrian/odoo13/odoo/odoo/http.py", line 515, in response_wrap
    response = f(*args, **kw)
  File "/Users/adrian/odoo13/odoo/addons/web/controllers/main.py", line 1331, in call_button
    action = self._call_kw(model, method, args, kwargs)
  File "/Users/adrian/odoo13/odoo/addons/web/controllers/main.py", line 1319, in _call_kw
    return call_kw(request.env[model], method, args, kwargs)
  File "/Users/adrian/odoo13/odoo/odoo/api.py", line 387, in call_kw
    result = _call_kw_multi(method, model, args, kwargs)
  File "/Users/adrian/odoo13/odoo/odoo/api.py", line 374, in _call_kw_multi
    result = method(recs, *args, **kwargs)
  File "<decorator-gen-56>", line 2, in button_immediate_install
  File "/Users/adrian/odoo13/odoo/odoo/addons/base/models/ir_module.py", line 72, in check_and_log
    return method(self, *args, **kwargs)
  File "/Users/adrian/odoo13/odoo/odoo/addons/base/models/ir_module.py", line 463, in button_immediate_install
    return self._button_immediate_function(type(self).button_install)
  File "/Users/adrian/odoo13/odoo/odoo/addons/base/models/ir_module.py", line 573, in _button_immediate_function
    modules.registry.Registry.new(self._cr.dbname, update_module=True)
  File "/Users/adrian/odoo13/odoo/odoo/modules/registry.py", line 86, in new
    odoo.modules.load_modules(registry._db, force_demo, status, update_module)
  File "/Users/adrian/odoo13/odoo/odoo/modules/loading.py", line 421, in load_modules
    processed_modules += load_marked_modules(cr, graph,
  File "/Users/adrian/odoo13/odoo/odoo/modules/loading.py", line 313, in load_marked_modules
    loaded, processed = load_module_graph(
  File "/Users/adrian/odoo13/odoo/odoo/modules/loading.py", line 202, in load_module_graph
    registry.init_models(cr, model_names, {'module': package.name}, new_install)
  File "/Users/adrian/odoo13/odoo/odoo/modules/registry.py", line 369, in init_models
    model._auto_init()
  File "/Users/adrian/odoo13/odoo/odoo/models.py", line 2524, in _auto_init
    new = field.update_db(self, columns)
  File "/Users/adrian/odoo13/odoo/odoo/fields.py", line 2456, in update_db
    return super(Many2one, self).update_db(model, columns)
  File "/Users/adrian/odoo13/odoo/odoo/fields.py", line 857, in update_db
    self.update_db_notnull(model, column)
  File "/Users/adrian/odoo13/odoo/odoo/fields.py", line 897, in update_db_notnull
    model._init_column(self.name)
  File "/Users/adrian/odoo13/odoo/odoo/models.py", line 2450, in _init_column
    value = field.default(self)
  File "/Users/adrian/odoo13/addons/test_install_order/models/website.py", line 16, in <lambda>
    custom_tag_id = fields.Many2one('product.custom.tag', default=lambda self: self.env.ref('test_install_order.default_tag'))
  File "/Users/adrian/odoo13/odoo/odoo/api.py", line 501, in ref
    return self['ir.model.data'].xmlid_to_object(xml_id, raise_if_not_found=raise_if_not_found)
  File "/Users/adrian/odoo13/odoo/odoo/addons/base/models/ir_model.py", line 1701, in xmlid_to_object
    t = self.xmlid_to_res_model_res_id(xmlid, raise_if_not_found)
  File "/Users/adrian/odoo13/odoo/odoo/addons/base/models/ir_model.py", line 1685, in xmlid_to_res_model_res_id
    return self.xmlid_lookup(xmlid)[1:3]
  File "<decorator-gen-25>", line 2, in xmlid_lookup
  File "/Users/adrian/odoo13/odoo/odoo/tools/cache.py", line 90, in lookup
    value = d[key] = self.method(*args, **kwargs)
  File "/Users/adrian/odoo13/odoo/odoo/addons/base/models/ir_model.py", line 1674, in xmlid_lookup
    raise ValueError('External ID not found in the system: %s' % xmlid)
ValueError: External ID not found in the system: test_install_order.default_tag
```

My guess here is that Odoo tries to build the default value (and therefore attempt to evaluate the line `self.env.ref('test_install_order.default_tag')` _before_ the data xml is evaluated.

My current work around is to have the default value commented out, install the addon (so that we have the external ID in database), then uncomment the default value and update the addon, and finally datafix the existing products for the default value using SQL.

### Expected Result

The addon is installed without any workaround required

**Comment by odoochina-dev** [comment]
When you wrote `default=lambda self: self.env.ref('test_install_order.default_tag')`, odoo is try to build a function (wrap the lambda) to be used as the field default getter  (`function address is needed!`), at this moment, the module is not installed and `self.env.ref('test_install_order.default_tag')` does not exists.

Try to define the field in another way: use default=_get_default_tag, where  _get_default_tag is a @api.model method, 
then the `self.env.ref('test_install_order.default_tag')` in this  _get_default_tag @api.model method will be delay to runtime and thus no problem.

**Comment by AdrianTeng** [comment]
@odoochina-dev I changed the model definition to as follow:

```
class ProductTemplate(models.Model):
    _inherit = 'product.template'

    @api.model
    def _get_default_tag(self):
        return self.env.ref('test_install_order.default_tag')

    custom_tag_id = fields.Many2one('product.custom.tag', default=_get_default_tag)
```

Sadly it still has the same problem -- namely the default function is evaluated before the xml insertion

**Comment by odoochina-dev** [comment]
@AdrianTeng It's quite strange, are you sure still the same problem? (odoo itself have lots of examples use default=_get_xxx, which returns dynamic values that only can decided at runtime. )

**Comment by AdrianTeng** [comment]
> @AdrianTeng It's quite strange, are you sure still the same problem? (odoo itself have lots of examples use default=_get_xxx, which returns dynamic values that only can decided at runtime. )

@odoochina-dev The problem here is slightly more complicated than just doing `default=_get_xxx` -- namely the get function returns a NEW MODEL within the same module.

I think when odoo install a module, it does all the table structure related changes (including setting default values) before handling the xml data

**Comment by OskOdoo** [comment]
Add this to resolve it for any forther same problem: self.env.ref('lims.default_rang_factor',False) because the installation starts with building tables structure than filling data and in this case the external ID is still not exist, so we replace it by False.

**Comment by OskOdoo** [comment]
Add this to resolve it for any forther same problem: self.env.ref('module.default_external_id',False) because the installation starts with building tables structure than filling data and in this case the external ID is still not exist, so we replace it by False.

**Comment by Yenthe666** [comment]
Or use installation/deinstallation hooks to make sure you have things in place before other logic goes off :)
